lib/galaxy/webapps/galaxy/services/history_contents.py

```diff
@@ -1602,11 +1602,13 @@ def _delete(self, item: HistoryItemModel):
 
     def _undelete(self, item: HistoryItemModel):
         if getattr(item, "purged", False):
-            return
+            raise exceptions.ItemDeletionException("This item has been permanently deleted and cannot be recovered.")
         manager = self._get_item_manager(item)
         manager.undelete(item, flush=self.flush)
 
     def _purge(self, item: HistoryItemModel, trans: ProvidesHistoryContext):
+        if getattr(item, "purged", False):
+            raise exceptions.ItemDeletionException("This item was already permanently deleted.")
```

Comment:
Maybe this can just be a no-op?
```suggestion
            return
```

Reply:
Sounds like we need to track the progress of a batch operation, it seems unfortunate that we have to raise an exception here to work around not having a way to do this currently. Could you just touch the HDA update_time for now ? That should also satisfy the update_time filter

Reply:
By the way, is it ok to raise the error when trying to `undelete` a `purged` dataset or should we do the same, no-op, and just touch the `update_time`?

Reply:
No, I'd raise an exception there.

Reply:
Done! Thanks for the reviews!